agent/src/main/java/org/openremote/agent/protocol/http/HTTPProtocol.java

```diff
@@ -602,4 +613,68 @@ protected void cancelPolling(AttributeRef attributeRef) {
         }
     }
 
+	public static class DynamicTimeReplacementClientRequestFilter implements ClientRequestFilter {
+
+		private final TimerService timerService;
+
+		private DynamicTimeReplacementClientRequestFilter(TimerService timerService) {
+			this.timerService = timerService;
+		}
+		@Override
+		public void filter(ClientRequestContext requestContext) throws IOException {
+			if (requestContext.hasEntity()) {
+				Object entity = requestContext.getEntity();
+
+				if (entity instanceof InputStream) {
+					String body = new Scanner((InputStream) entity, StandardCharsets.UTF_8).useDelimiter("\\A").next();
```

Comment:
injection of dynamic values should be done in the protocol as body content may not be text.

Also injection must be supported on the query params and maybe headers which is likely the more common requirement rather than in the body (think adding date range for a query)

Reply:
What exactly do you mean by "In the protocol"? I was not exactly sure where to inject the filter so I thought of doing it this way. What I can do is ensure that replacement happens for `requestContext.getEntity()`, `requestContext.getHeaders()`, and `requestContext.getProperty(QUERY_PARAMETERS_PROPERTY)`, to make sure that all of them are checked.

Reply:
OK to clarify:

For body injection see `ProtocolUtil.doOutboundValueProcessing()` which is the generic protocol mechanism for formatting the body for outbound protocol requests.

For query parameter injection see `QueryParameterInjectorFilter.java`

For path injection see `HTTPProtocol.getRequestBuilder()`

Reply:
The logic should be reused in these places so a static public method will be needed